Fix channel bans and user silences not being checked against account host when the user has a sethost.

Backport from snircd 1.4.0 - http://hg.quakenet.org/snircd/rev/876732831347

diff -r 98f03fd09b81 ircd/channel.c
--- a/ircd/channel.c
+++ b/ircd/channel.c
@@ -374,22 +374,25 @@
   char        tmphost[HOSTLEN + 1];
   char        iphost[SOCKIPLEN + 1];
   char       *hostmask;
-  char       *sr;
+  char       *sr = NULL; /* realhost */
+  char       *sa = NULL; /* account host */
   struct Ban *found;
 
   /* Build nick!user and alternate host names. */
   ircd_snprintf(0, nu, sizeof(nu), "%s!%s",
                 cli_name(cptr), cli_user(cptr)->username);
   ircd_ntoa_r(iphost, &cli_ip(cptr));
-  if (!IsAccount(cptr))
-    sr = NULL;
-  else if (HasHiddenHost(cptr) || HasSetHost(cptr))
+
+  /* user has a hiddenhost (+rx) or has a sethost (+h) - check realhost */
+  if (HasHiddenHost(cptr) || IsSetHost(cptr))
     sr = cli_user(cptr)->realhost;
-  else
-  {
+
+  /* user has an account (+r),
+   *   but does not have a hiddenhost (-x) or has a sethost (+h) - check account host */
+  if (IsAccount(cptr) && (!IsHiddenHost(cptr) || IsSetHost(cptr))) {
     ircd_snprintf(0, tmphost, HOSTLEN, "%s.%s",
                   cli_user(cptr)->account, feature_str(FEAT_HIDDEN_HOST));
-    sr = tmphost;
+    sa = tmphost;
   }
 
   /* Walk through ban list. */
@@ -409,7 +412,8 @@
     if (!((banlist->flags & BAN_IPMASK)
          && ipmask_check(&cli_ip(cptr), &banlist->address, banlist->addrbits))
         && match(hostmask, cli_user(cptr)->host)
-        && !(sr && !match(hostmask, sr)))
+        && !(sr && !match(hostmask, sr))
+        && !(sa && !match(hostmask, sa)))
         continue;
     /* If an exception matches, no ban can match. */
     if (banlist->flags & BAN_EXCEPTION)
